chainer/functions/math/basic_math.py

```diff
@@ -430,29 +431,76 @@ def check_type_forward(self, in_types):
             in_types[0].shape == in_types[1].shape
         )
 
-    def forward_cpu(self, x):
-        self.y = utils.force_array(x[0] ** x[1])
+    def forward(self, x):
+        self.retain_inputs((0, 1))
+        self.y = x[0] ** x[1]
+
+        if isinstance(x[0], numpy.ndarray):
+            return utils.force_array(self.y),
         return self.y,
 
-    def forward_gpu(self, x):
-        return x[0] ** x[1],
+    def backward(self, indexes, gy):
+        inputs = self.get_retained_inputs()
+        return PowVarVarGrad(self.y).apply(inputs + gy)
 
-    def backward_cpu(self, x, gy):
-        one = x[1].dtype.type(1)
-        gx0 = utils.force_array(x[1] * (x[0] ** (x[1] - one)) * gy[0])
-        gx1 = utils.force_array(numpy.log(x[0]) * self.y * gy[0])
+
+class PowVarVarGrad(function_node.FunctionNode):
+
+    def __init__(self, y=None):
+        self.y = y
+
+    def check_type_forward(self, in_types):
+        type_check.expect(in_types.size() == 3)
+        type_check.expect(
+            in_types[0].dtype.kind == 'f',
+            in_types[0].dtype == in_types[1].dtype,
+            in_types[0].shape == in_types[1].shape,
+            in_types[0].dtype == in_types[2].dtype,
+            in_types[0].shape == in_types[2].shape,
+        )
+
+    def forward(self, inputs):
+        self.retain_inputs((0, 1, 2))
+        x0, x1, gy = inputs
+
+        if isinstance(x0, numpy.ndarray):
+            one = x1.dtype.type(1)
+            gx0 = utils.force_array(x1 * (x0 ** (x1 - one)) * gy)
+            gx1 = utils.force_array(numpy.log(x0) * self.y * gy)
+        else:
+            gx0, gx1 = cuda.elementwise(
+                'T x0, T x1, T gy, T y', 'T gx0, T gx1',
+                '''
+                gx0 = x1 * pow(x0, x1 - 1) * gy;
+                gx1 = log(x0) * y * gy;
+                ''', 'pow_var_var_bwd')(x0, x1, gy, self.y)
         return gx0, gx1
 
-    def backward_gpu(self, x, gy):
-        return cuda.elementwise(
-            'T x0, T x1, T gy', 'T gx0, T gx1',
-            '''
-               gx0 = x1 * pow(x0, x1 - 1) * gy;
-               gx1 = log(x0) * pow(x0, x1) * gy;
-            ''', 'pow_var_var_bwd')(x[0], x[1], gy[0])
+    def backward(self, indexes, ggx):
+        x0, x1, gy = self.get_retained_inputs()
+        ggx0, ggx1 = ggx
+
+        log_x0 = chainer.functions.log(x0)
+        pow_x0_x1 = x0 ** x1
+        pow_x0_x1_1 = x0 ** (x1 - 1)
+        pow_x0_x1_2 = x0 ** (x1 - 2)
+
+        ret = []
+        if 0 in indexes:
+            gx0 = (ggx0 * x1 * (x1 - 1) * pow_x0_x1_2 +
+                   ggx1 * (pow_x0_x1_1 * (log_x0 * x1 + 1))) * gy
```

Comment:
Would you remove the outer parentheses of `(pow_x0_x1_1 * (log_x0 * x1 + 1))`?

Reply:
Fixed.